File: daily_problems/2025/04/0424/personal_submission/cf234h_yawn_sean.py
# Submission link: https://codeforces.com/contest/234/submission/316896239
def main():
    input_file = open("input.txt", 'r')
    input = lambda: input_file.readline().strip()

    output_file = open("output.txt", 'w')
    def print(s):
        output_file.write(s)
        output_file.write('\n')

    n1 = II()
    nums1 = LII()
    n2 = II()
    nums2 = LII()

    nums1.append(0)
    nums2.append(0)

    c1 = 0
    for i in range(n1):
        if nums1[i] != nums1[i + 1]:
            c1 += 1

    c2 = 0
    for i in range(n2):
        if nums2[i] != nums2[i + 1]:
            c2 += 1

    if c1 < c2:
        order = []
        vals = []
        
        p = 0
        for i in range(n2):
            while p < n1 and nums1[p] == nums2[i]:
                order.append(p)
                vals.append(nums1[p])
                p += 1
            order.append(n1 + i)
            vals.append(nums2[i])
        
        for i in range(p, n1):
            order.append(i)
            vals.append(nums1[i])

    else:
        order = []
        vals = []
        
        p = 0
        for i in range(n1):
            while p < n2 and nums2[p] == nums1[i]:
                order.append(n1 + p)
                vals.append(nums2[p])
                p += 1
            order.append(i)
            vals.append(nums1[i])
        
        for i in range(p, n2):
            order.append(n1 + i)
            vals.append(nums2[i])

    print(' '.join(str(x + 1) for x in order))

    vals.append(0)
    idxs = [i for i in range(1, n1 + n2 + 1) if vals[i] != vals[i - 1]]
    print(str(len(idxs)))
    print(' '.join(map(str, idxs)))

File: daily_problems/2025/04/0424/personal_submission/cf234h_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/convolution"
// #include "atcoder/dsu"
// #include "atcoder/lazysegtree"
// #include "atcoder/fenwicktree"
// #include "atcoder/math"
// #include "atcoder/maxflow"
// #include "atcoder/mincostflow"
// #include "atcoder/modint"
// #include "atcoder/scc"
// #include "atcoder/segtree"
// #include "atcoder/string"
// #include "atcoder/twosat"
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divided, I conquered!

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    ifstream fin("input.txt");
    ofstream fout("output.txt");

    int n1;
    fin >> n1;

    vector<int> nums1(n1);
    for (auto &v: nums1) fin >> v;

    int n2;
    fin >> n2;

    vector<int> nums2(n2);
    for (auto &v: nums2) fin >> v;

    nums1.emplace_back(0);
    nums2.emplace_back(0);

    int c1 = 0, c2 = 0;

    for (int i = 1; i <= n1; i ++) {
        if (nums1[i] != nums1[i - 1]) {
            c1 ++;
        }
    }
    
    for (int i = 1; i <= n2; i ++) {
        if (nums2[i] != nums2[i - 1]) {
            c2 ++;
        }
    }

    vector<int> vals;
    int p = 0;

    if (c1 < c2) {
        for (int i = 0; i < n2; i ++) {
            while (p < n1 && nums1[p] == nums2[i]) {
                fout << p + 1 << ' ';
                vals.emplace_back(nums1[p]);
                p ++;
            }
            fout << n1 + i + 1 << ' ';
            vals.emplace_back(nums2[i]);
        }

        while (p < n1) {
            fout << p + 1 << ' ';
            vals.emplace_back(nums1[p]);
            p ++;
        }
    }

    else {
        for (int i = 0; i < n1; i ++) {
            while (p < n2 && nums2[p] == nums1[i]) {
                fout << n1 + p + 1 << ' ';
                vals.emplace_back(nums2[p]);
                p ++;
            }
            fout << i + 1 << ' ';
            vals.emplace_back(nums1[i]);
        }

        while (p < n2) {
            fout << n1 + p + 1 << ' ';
            vals.emplace_back(nums2[p]);
            p ++;
        }
    }

    fout << '\n';

    vals.emplace_back(0);
    vector<int> idxs;

    for (int i = 1; i <= n1 + n2; i ++) {
        if (vals[i] != vals[i - 1]) {
            idxs.emplace_back(i);
        }
    }

    fout << idxs.size() << '\n';
    for (auto &idx: idxs) fout << idx << ' ';

    return 0;
}